### ISSUE TITLE
Fused SGD on CPU silently skips vectorized fp16 and bf16 parameter updates in certain circumstances

### ISSUE THREAD

**Issue by myfykris** [pytorch/pytorch#191763]
### 🐛 Describe the bug

The optimizer `torch.optim.SGD` fails to update the vectorized parts of fp16 and bf16 params when numel() >= SIMD width (SIMD width = `lpVec::size()` / `at::vec::Vectorized<scalar_t>::size()` in the CPU fused-SGD kernel, and that == 8 on my system) and fused=True.  Works fine for fp32 and fp64, only the low precision types have the problem.

- when numel() < SIMD width, the angels sing and world peace begins because it works
- when numel() % SIMD width == 0, the param update is a no-op
- when numel() > SIMD width, every complete SIMD block isn't touched, and only the scalar remainder updates

PyTorch internal testing does not find this bug because it's model has small params, sizes 6, 3, 3, and 1, and that isn't going to be >= SIMD width on any hardware I know of, and the existing large tensor fused optimizer test uses zero gradients so the result is expected to not have changed params, and hence doesn't expose the issue.

## Reproducing the issue

I reproduced this on my m3 max with both PyTorch 2.12.1 and 2.13.0.  

Code to reproduce:
```python
import platform
import torch

def run_step(dtype: torch.dtype, size: int, *, fused: bool) -> torch.Tensor:
    parameter = torch.nn.Parameter(torch.ones(size, dtype=dtype, device="cpu"))
    parameter.grad = torch.ones_like(parameter)

    torch.optim.SGD([parameter], lr=0.125, fused=fused).step()
    return parameter.detach()


def result_text(tensor: torch.Tensor) -> str:
    total = tensor.numel()
    updated = int(torch.count_nonzero(tensor == 0.875).item())
    return f"{updated} out of {total} elements updated to 0.875"


print(f"PyTorch: {torch.__version__}")
print(f"Platform: {platform.platform()}")

for dtype in (torch.float16, torch.bfloat16):
    for size in (7, 8, 9, 64, 65):
        unfused = run_step(dtype, size, fused=False)
        fused = run_step(dtype, size, fused=True)

        print(f"\n{dtype}, {size} elements")
        print(f"  unfused: {result_text(unfused)}")
        print(f"  fused:   {result_text(fused)}")
```

Output for 2.13.0
```text
PyTorch: 2.13.0
Platform: macOS-26.3-arm64-arm-64bit-Mach-O

torch.float16, 7 elements
  unfused: 7 out of 7 elements updated to 0.875
  fused:   7 out of 7 elements updated to 0.875

torch.float16, 8 elements
  unfused: 8 out of 8 elements updated to 0.875
  fused:   0 out of 8 elements updated to 0.875

torch.float16, 9 elements
  unfused: 9 out of 9 elements updated to 0.875
  fused:   1 out of 9 elements updated to 0.875

torch.float16, 64 elements
  unfused: 64 out of 64 elements updated to 0.875
  fused:   0 out of 64 elements updated to 0.875

torch.float16, 65 elements
  unfused: 65 out of 65 elements updated to 0.875
  fused:   1 out of 65 elements updated to 0.875

torch.bfloat16, 7 elements
  unfused: 7 out of 7 elements updated to 0.875
  fused:   7 out of 7 elements updated to 0.875

torch.bfloat16, 8 elements
  unfused: 8 out of 8 elements updated to 0.875
  fused:   0 out of 8 elements updated to 0.875

torch.bfloat16, 9 elements
  unfused: 9 out of 9 elements updated to 0.875
  fused:   1 out of 9 elements updated to 0.875

torch.bfloat16, 64 elements
  unfused: 64 out of 64 elements updated to 0.875
  fused:   0 out of 64 elements updated to 0.875

torch.bfloat16, 65 elements
  unfused: 65 out of 65 elements updated to 0.875
  fused:   1 out of 65 elements updated to 0.875
```

Output for 2.12.1
```text
PyTorch: 2.12.1
Platform: macOS-26.3-arm64-arm-64bit-Mach-O

torch.float16, 7 elements
  unfused: 7 out of 7 elements updated to 0.875
  fused:   7 out of 7 elements updated to 0.875

torch.float16, 8 elements
  unfused: 8 out of 8 elements updated to 0.875
  fused:   0 out of 8 elements updated to 0.875

torch.float16, 9 elements
  unfused: 9 out of 9 elements updated to 0.875
  fused:   1 out of 9 elements updated to 0.875

torch.float16, 64 elements
  unfused: 64 out of 64 elements updated to 0.875
  fused:   0 out of 64 elements updated to 0.875

torch.float16, 65 elements
  unfused: 65 out of 65 elements updated to 0.875
  fused:   1 out of 65 elements updated to 0.875

torch.bfloat16, 7 elements
  unfused: 7 out of 7 elements updated to 0.875
  fused:   7 out of 7 elements updated to 0.875

torch.bfloat16, 8 elements
  unfused: 8 out of 8 elements updated to 0.875
  fused:   0 out of 8 elements updated to 0.875

torch.bfloat16, 9 elements
  unfused: 9 out of 9 elements updated to 0.875
  fused:   1 out of 9 elements updated to 0.875

torch.bfloat16, 64 elements
  unfused: 64 out of 64 elements updated to 0.875
  fused:   0 out of 64 elements updated to 0.875

torch.bfloat16, 65 elements
  unfused: 65 out of 65 elements updated to 0.875
  fused:   1 out of 65 elements updated to 0.875
```

### Cause

In the fused SGD kernel for fp16 and bf16 the vectorized loop does the calculations it needs, but then the loop ends without applying the LR update.  Like any sane developer now days I use AI so I can be lazy, so here's what AI put together to point out where the problem is:

## Source of the problem

In the FP16/BF16 specialization of the CPU fused-SGD kernel, the
vectorized loop calculates the gradients and momentum but reaches the
end of the loop without applying the learning-rate update to
`param_vec1` and `param_vec2` or storing them back to `param_ptr`:

https://github.com/pytorch/pytorch/blob/4a4f3e8cfe754f198669b258a5e1c3f8cec0f247/aten/src/ATen/native/cpu/FusedSGDKernel.cpp#L35-L74

The scalar remainder correctly updates `param_ptr`:

https://github.com/pytorch/pytorch/blob/4a4f3e8cfe754f198669b258a5e1c3f8cec0f247/aten/src/ATen/native/cpu/FusedSGDKernel.cpp#L75-L102

The FP32/FP64 vectorized implementation also performs and stores its
parameter update:

https://github.com/pytorch/pytorch/blob/4a4f3e8cfe754f198669b258a5e1c3f8cec0f247/aten/src/ATen/native/cpu/FusedSGDKernel.cpp#L126-L156

The same missing update is still present in the current 2.14
development source. 

### Versions

PyTorch version: 2.12.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.3 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.6.3.2)
CMake version: version 4.3.4
Libc version: N/A

Python version: 3.14.2 (main, Dec  5 2025, 16:49:16) [Clang 17.0.0 (clang-1700.4.4.1)] (64-bit runtime)
Python platform: macOS-26.3-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] numpy==2.5.0
[pip3] torch==2.12.1
[pip3] torchcts==0.4.1
[conda] numpy                     2.5.1                    pypi_0    pypi
[conda] numpydoc                  1.7.0           py312hca03da5_0  
[conda] tbb                       2021.8.0             h48ca7d4_0  
[conda] torch                     2.6.0                    pypi_0    pypi

cc @ezyang @gchanan @kadeng @msaroufim @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

**Comment by janeyx99** [comment]
Yikes, thanks for spotting this, I would happily review a PR fixing this with a test

**Comment by myfykris** [comment]
> Yikes, thanks for spotting this, I would happily review a PR fixing this with a test

If I do the 2.14 will you guys handle inclusion in the 2.13 patch in August?

**Comment by PrabinDevkota** [comment]
Hi @janeyx99,

I’ve opened a PR here: <https://github.com/pytorch/pytorch/pull/191774>
**Changes:**
- In the fused CPU SGD Half/BFloat16 vectorized loop, apply the learning-rate update and store the params (this was previously missing, so full SIMD blocks were left unchanged).
- Fix momentum-buffer reloads to use `lpVec::loadu` + `convert_to_float` (instead of `fVec::loadu` through a Half/BF16 pointer), so multi-step momentum stays correct once params update.
- Add a CPU regression test comparing fused vs unfused SGD for `float16`/`bfloat16` with large enough `numel` to exercise the vectorized path (including a second step when momentum is used).

Would you mind taking a look when you have a chance? Happy to address any feedback. Thanks!

**Comment by janeyx99** [comment]
@myfykris We don't backport fixes, so it won't be in 2.13, but the 2.14 release should not be too far from now.

**Comment by myfykris** [comment]
> Hi [@janeyx99](https://github.com/janeyx99),
> 
> I’ve opened a PR here: [#191774](https://github.com/pytorch/pytorch/pull/191774) **Changes:**
> 
> * In the fused CPU SGD Half/BFloat16 vectorized loop, apply the learning-rate update and store the params (this was previously missing, so full SIMD blocks were left unchanged).
> * Fix momentum-buffer reloads to use `lpVec::loadu` + `convert_to_float` (instead of `fVec::loadu` through a Half/BF16 pointer), so multi-step momentum stays correct once params update.
> * Add a CPU regression test comparing fused vs unfused SGD for `float16`/`bfloat16` with large enough `numel` to exercise the vectorized path (including a second step when momentum is used).
> 
> Would you mind taking a look when you have a chance? Happy to address any feedback. Thanks!

Nice sniping.

**Comment by myfykris** [comment]
> [@myfykris](https://github.com/myfykris) We don't backport fixes, so it won't be in 2.13, but the 2.14 release should not be too far from now.

Sorry, I thought you did when it was a correctness issue and you hadn't released the patch version yet, my bad.

**Comment by janeyx99** [comment]
Oh like a cherry-pick, yes, we do have those, but even then the fix would have to fix something added in the release, which fusedsgd on cpu predates.

**Comment by myfykris** [comment]
> Oh like a cherry-pick, yes, we do have those, but even then the fix would have to fix something added in the release, which fusedsgd on cpu predates.

Ok, thanks for explaining the policy and your quick responses.  Have a good day!
